File: src/main/java/legacy/PathButton.java
package legacy;

import java.awt.Component;
import java.awt.Dialog;
import java.awt.FileDialog;
import java.awt.Frame;
import java.awt.datatransfer.DataFlavor;
import java.awt.datatransfer.Transferable;
import java.awt.datatransfer.UnsupportedFlavorException;
import java.awt.event.MouseEvent;
import java.io.File;
import java.io.FilenameFilter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import javax.swing.JComponent;
import javax.swing.SwingUtilities;
import javax.swing.TransferHandler;
import javax.swing.event.MouseInputAdapter;

import net.roydesign.ui.FolderDialog;

/**
 *  This class is a rewritten version
 *  of FScape's <code>PathIcon</code> and provides
 *  a simple ToolIcon like button to
 *  allow the user to select a file
 *  from the harddisk. Besides, the user
 *  can drag files from the Finder onto
 *  the button's icon to set the button's
 *  path.
 *
 *  @author		Hanns Holger Rutz
 *  @version	0.18, 07-Feb-10
 *
 *  @see		java.awt.FileDialog
 *  @see		net.roydesign.ui.FolderDialog
 */
public class PathButton
extends ModificationButton
implements EventManager.Processor
{
	private File				path	= null;
	private final int			type;
	private String				dlgTxt;
	private final EventManager	elm		= new EventManager( this );
	private FilenameFilter		filter	= null;

	protected static final DataFlavor[] supportedFlavors = {
		DataFlavor.javaFileListFlavor, DataFlavor.stringFlavor
	};

	public PathButton()
	{
		this( PathField.TYPE_INPUTFILE );
	}

	public PathButton( int type )
	{
		super( SHAPE_LIST );
		this.type   = type;

		setToolTipText( GUIUtil.getResourceString( "buttonChoosePathTT" ));
		setTransferHandler( new PathTransferHandler() );

		final MouseInputAdapter mia = new MouseInputAdapter() {
			private MouseEvent dndInit = null;
			private boolean dndStarted = false;

			public void mousePressed( MouseEvent e )
			{
				dndInit		= e;
				dndStarted	= false;
			}

			public void mouseReleased( MouseEvent e )
			{
				if( !dndStarted && contains( e.getPoint() )) showFileChooser();
				dndInit		= null;
				dndStarted	= false;
			}

			public void mouseDragged( MouseEvent e )
			{
				if( !dndStarted && (dndInit != null) &&
					((Math.abs( e.getX() - dndInit.getX() ) > 5) ||
					 (Math.abs( e.getY() - dndInit.getY() ) > 5))) {

					JComponent c = (JComponent) e.getSource();
					c.getTransferHandler().exportAsDrag( c, e, TransferHandler.COPY );
					dndStarted = true;
				}
			}
		};

		addMouseListener( mia );
		addMouseMotionListener( mia );
	}

	/**
	 *  Constructs a new <code>PathButton</code> for
	 *  given type of file chooser and optional dialog text
	 *
	 *  @param  type	the type of file chooser to display. the values
	 *					are those from <code>PathField</code>, e.g.
	 *					<code>PathField.TYPE_INPUT</code>
	 *  @param  dlgTxt  text to display in the file chooser dialog or <code>null</code>
	 */
	public PathButton( int type, String dlgTxt )
	{
		this( type );
		setDialogText( dlgTxt );
	}

	public void setDialogText( String dlgTxt )
	{
		this.dlgTxt = dlgTxt;
	}

	/**
	 *  Sets the button's path. This is path will be
	 *  used as default setting when the file chooser is shown
	 *
	 *  @param  path	the new path for the button
	 */
	public void setPath( File path )
	{
		this.path = path;
	}

	/*
	 *  Sets a new path and dispatches a <code>PathEvent</code>
	 *  to registered listeners
	 *
	 *  @param  path	the new path for the button and the event
	 */
	protected void setPathAndDispatchEvent( File path )
	{
		setPath( path );
		elm.dispatchEvent( new PathEvent( this, PathEvent.CHANGED, System.currentTimeMillis(), path ));
	}

	/**
	 *  Returns the path set for the button
	 *  or chosen by the user after a file chooser
	 *  has been shown.
	 *
	 *  @return the button's path or <code>null</code>
	 *			if no path was set or the file chooser was cancelled
	 */
	public File getPath()
	{
		return path;
	}

	/**
	 * 	Sets the filter to use for enabling or disabling items
	 * 	in the file dialog.
	 *
	 * 	@param	filter	the new filter or null to remove an existing filter
	 */
	public void setFilter( FilenameFilter filter )
	{
		this.filter = filter;
	}

	/**
	 * 	Queries the current item filter for the file dialog.
	 *
	 * 	@return	the current filter or null if no filter is installed
	 */
	public FilenameFilter getFilter()
	{
		return filter;
	}

	// --- listener registration ---

	/**
	 *  Register a <code>PathListener</code>
	 *  which will be informed about changes of
	 *  the path (i.e. user selections in the
	 *  file chooser).
	 *
	 *  @param  listener	the <code>PathListener</code> to register
	 *  @see	de.sciss.app.EventManager#addListener( Object )
	 */
	public void addPathListener( PathListener listener )
	{
		elm.addListener( listener );
	}

	/**
	 *  Unregister a <code>PathListener</code>
	 *  from receiving path change events.
	 *
	 *  @param  listener	the <code>PathListener</code> to unregister
	 *  @see	de.sciss.app.EventManager#removeListener( Object )
	 */
	public void removePathListener( PathListener listener )
	{
		elm.removeListener( listener );
	}

	public void processEvent( BasicEvent e )
	{
		PathListener listener;
		int i;

		for( i = 0; i < elm.countListeners(); i++ ) {
			listener = (PathListener) elm.getListener( i );
			switch( e.getID() ) {
			case PathEvent.CHANGED:
				listener.pathChanged( (PathEvent) e );
				break;
			default:
				assert false : e.getID();
			}
		} // for( i = 0; i < elm.countListeners(); i++ )
	}

	protected void showDialog( Dialog dlg )
	{
		dlg.setVisible( true );
	}

	protected void showFileChooser()
	{
		File		p;
		FileDialog	fDlg;
		String		fDir, fFile; // , fPath;
//		int			i;
		Component	win;

		for( win = this; !(win instanceof Frame); ) {
			win = SwingUtilities.getWindowAncestor( win );
			if( win == null ) return;
		}

		p = getPath();
		switch( type & PathField.TYPE_BASICMASK ) {
		case PathField.TYPE_INPUTFILE:
			fDlg = new FileDialog( (Frame) win, dlgTxt, FileDialog.LOAD );
			break;
		case PathField.TYPE_OUTPUTFILE:
			fDlg = new FileDialog( (Frame) win, dlgTxt, FileDialog.SAVE );
			break;
		case PathField.TYPE_FOLDER:
			fDlg = new FolderDialog( (Frame) win, dlgTxt );
			break;
		default:
			fDlg = null;
			assert false : (type & PathField.TYPE_BASICMASK);
			break;
		}
		if( p != null ) {
			fDlg.setFile( p.getName() );
			fDlg.setDirectory( p.getParent() );
		}
		if( filter != null ) {
			fDlg.setFilenameFilter( filter );
		}
		showDialog( fDlg );
		fDir	= fDlg.getDirectory();
		fFile	= fDlg.getFile();

		if( ((type & PathField.TYPE_BASICMASK) != PathField.TYPE_FOLDER) && (fDir == null) ) {
			fDir = "";
		}

		if( (fFile != null) && (fDir != null) ) {

			if( (type & PathField.TYPE_BASICMASK) == PathField.TYPE_FOLDER ) {
				p = new File( fDir );
			} else {
				p = new File( fDir + fFile );
			}
			setPathAndDispatchEvent( p );
		}

		fDlg.dispose();
	}

// ----------- interner TransferHandler -----------

	private class PathTransferHandler
	extends TransferHandler
	{
		protected PathTransferHandler() { /* empty */ }

		/**
		 * Overridden to import a Pathname (Fileliste or String) if it is available.
		 */
		public boolean importData( JComponent c, Transferable t )
		{
			Object		o;
			List		fileList;
			File		newPath	= null;

			try {
				if( t.isDataFlavorSupported( DataFlavor.javaFileListFlavor )) {
					o =  t.getTransferData( DataFlavor.javaFileListFlavor );
					if( o instanceof List ) {
						fileList = (List) o;
						if( !fileList.isEmpty() ) {
							o  =  fileList.get( 0 );
							if( o instanceof File ) {
								newPath = (File) o;
							} else {
								newPath = new File( o.toString() );
							}
						}
					}
				} else if( t.isDataFlavorSupported( DataFlavor.stringFlavor )) {
					newPath = new File( (String) t.getTransferData( DataFlavor.stringFlavor ));
				}
				if( newPath != null ) {
					setPathAndDispatchEvent( newPath );
					return true;
				}
			}
			catch( UnsupportedFlavorException e1 ) { e1.printStackTrace(); }
			catch( IOException e2 ) { e2.printStackTrace(); }

			return false;
		}

		public int getSourceActions( JComponent c )
		{
			return COPY;
		}

		protected Transferable createTransferable( JComponent c )
		{
//			System.err.println( "createTransferable" );
			return new PathTransferable( getPath() );
		}

		protected void exportDone( JComponent source, Transferable data, int action )
		{
//			System.err.println( "exportDone. Action == "+action );
		}

		public boolean canImport( JComponent c, DataFlavor[] flavors )
		{
// System.err.println( "canImport" );

			for( int i = 0; i < flavors.length; i++ ) {
				for( int j = 0; j < supportedFlavors.length; j++ ) {
					if( flavors[i].equals( supportedFlavors[j] )) return true;
				}
			}
			return false;
		}

//		public Icon getVisualRepresentation( Transferable t )
//		{
//System.err.println( "getVisualRepresentation" );
//			return FileSystemView.getFileSystemView().getSystemIcon( new File( System.getProperty( "user.home" )));
//		}
	} // class PathTransferHandler

	private static class PathTransferable
	implements Transferable
	{
		private final File f;

		protected PathTransferable( File f )
		{
			this.f	= f;
		}

		public DataFlavor[] getTransferDataFlavors()
		{
			return supportedFlavors;
		}

		public boolean isDataFlavorSupported( DataFlavor flavor )
		{
			for( int i = 0; i < supportedFlavors.length; i++ ) {
				if( supportedFlavors[ i ].equals( flavor )) return true;
			}
			return false;
		}

		public Object getTransferData( DataFlavor flavor )
		throws UnsupportedFlavorException, IOException
		{
			if( f == null ) throw new IOException();
			if( flavor.equals( DataFlavor.javaFileListFlavor )) {
				final List coll = new ArrayList( 1 );
				coll.add( f );
				return coll;
			} else if( flavor.equals( DataFlavor.stringFlavor )) {
				return f.getAbsolutePath();
			}
			throw new UnsupportedFlavorException( flavor );
		}
	}
}

File: src/main/scala/de/sciss/kontur/desktop/impl/PathButton.scala
package de.sciss.kontur
package desktop
package impl

import java.awt.datatransfer.{UnsupportedFlavorException, Transferable, DataFlavor}
import java.io.{FilenameFilter, FileFilter, IOException, File}
import legacy.{BasicEvent, PathListener, PathEvent, GUIUtil, EventManager, ModificationButton}
import javax.swing.event.MouseInputAdapter
import java.awt.event.MouseEvent
import javax.swing.{SwingUtilities, TransferHandler, JComponent}
import java.awt.{Component, Frame, FileDialog, Dialog}
import annotation.tailrec
import net.roydesign.ui.FolderDialog

/**
 * This class is a rewritten version
 * of FScape's <code>PathIcon</code> and provides
 * a simple ToolIcon like button to
 * allow the user to select a file
 * from the harddisk. Besides, the user
 * can drag files from the Finder onto
 * the button's icon to set the button's
 * path.
 */
object PathButton {
  private final val supportedFlavors = DataFlavor.javaFileListFlavor :: DataFlavor.stringFlavor :: Nil

  private class PathTransferable(f: File) extends Transferable {
    def getTransferDataFlavors: Array[DataFlavor] = supportedFlavors.toArray
    def isDataFlavorSupported(flavor: DataFlavor): Boolean = supportedFlavors.contains(flavor)

    def getTransferData(flavor: DataFlavor): AnyRef = {
      if (flavor == DataFlavor.javaFileListFlavor) {
        val coll: java.util.List[File] = new java.util.ArrayList[File](1)
        coll.add(f)
        coll
      }
      else if (flavor == DataFlavor.stringFlavor) {
        f.getAbsolutePath
      }
      throw new UnsupportedFlavorException(flavor)
    }
  }
}
class PathButton(var mode: PathField.Mode = PathField.Input)
  extends ModificationButton(ModificationButton.SHAPE_LIST) with EventManager.Processor {

  import PathButton._

  private final val elm: EventManager = new EventManager(this)

  setToolTipText(GUIUtil.getResourceString("buttonChoosePathTT"))
  setTransferHandler(new PathTransferHandler)

  var dialogText  = ""
  var file        = Option.empty[File]
  var filter      = Option.empty[FileFilter]

  private object mia extends MouseInputAdapter {
    private var dndInit     = Option.empty[MouseEvent]
    private var dndStarted  = false

    override def mousePressed(e: MouseEvent) {
      dndInit     = Some(e)
      dndStarted  = false
    }

    override def mouseReleased(e: MouseEvent) {
      if (!dndStarted && contains(e.getPoint)) showFileChooser()
      dndInit     = None
      dndStarted  = false
    }

    override def mouseDragged(e: MouseEvent) {
      dndInit.foreach { e0 =>
        if (!dndStarted && ((math.abs(e.getX - e0.getX) > 5) || (math.abs(e.getY - e0.getY) > 5))) {
          e.getSource match {
            case c: JComponent =>
              c.getTransferHandler.exportAsDrag(c, e, TransferHandler.COPY)
              dndStarted = true
            case _ =>
          }
        }
      }
    }
  }

  addMouseListener(mia)
  addMouseMotionListener(mia)

  private def setFileAndDispatchEvent(f: File) {
    file = Some(f)
    elm.dispatchEvent(new PathEvent(this, PathEvent.CHANGED, System.currentTimeMillis, f))
  }

  /**
   * Register a <code>PathListener</code>
   * which will be informed about changes of
   * the path (i.e. user selections in the
   * file chooser).
   *
   * @param  listener	the <code>PathListener</code> to register
   * @see	de.sciss.app.EventManager#addListener( Object )
   */
  def addPathListener(listener: PathListener) {
    elm.addListener(listener)
  }

  /**
   * Unregister a <code>PathListener</code>
   * from receiving path change events.
   *
   * @param  listener	the <code>PathListener</code> to unregister
   * @see	de.sciss.app.EventManager#removeListener( Object )
   */
  def removePathListener(listener: PathListener) {
    elm.removeListener(listener)
  }

  def processEvent(e: BasicEvent) {
    e match {
      case pe: PathEvent =>
        var i = 0
        while (i < elm.countListeners) {
          val listener = elm.getListener(i).asInstanceOf[PathListener]
          listener.pathChanged(pe)
          i += 1
        }
    }
  }

  protected def showDialog(dlg: Dialog) {
    dlg.setVisible(true)
  }

  private def showFileChooser() {
    @tailrec def findFrame(parent: Component): Frame = {
      SwingUtilities.getWindowAncestor(parent) match {
        case f: Frame => f
        case w if w != null => findFrame(w)
        case _ => null
      }
    }

    val win     = findFrame(this)
    val dlgTxt  = if (dialogText == "") null else dialogText
    val fDlg = mode match {
      case PathField.Input =>
        new FileDialog(win, dlgTxt, FileDialog.LOAD)
      case PathField.Output =>
        new FileDialog(win, dlgTxt, FileDialog.SAVE)
      case PathField.Folder =>
        new FolderDialog(win, dlgTxt)
    }
    file.foreach { f =>
      fDlg.setFile(f.getName)
      fDlg.setDirectory(f.getParent)
    }
    filter.foreach(flt => fDlg.setFilenameFilter(new FilenameFilter {
      def accept(dir: File, name: String) = flt.accept(new File(dir, name))
    }))
    showDialog(fDlg)
    val fDir  = fDlg.getDirectory
    val fFile = fDlg.getFile
    if ((fFile != null) && (fDir != null)) {
      val p = if (mode == PathField.Folder) {
        new File(fDir)
      } else {
        new File(fDir + fFile)
      }
      setFileAndDispatchEvent(p)
    }
    fDlg.dispose()
  }

  private class PathTransferHandler extends TransferHandler {
    /**
     * Overridden to import a Pathname (Fileliste or String) if it is available.
     */
    override def importData(c: JComponent, t: Transferable): Boolean = {
      val newPath = if (t.isDataFlavorSupported(DataFlavor.javaFileListFlavor)) {
        t.getTransferData(DataFlavor.javaFileListFlavor) match {
          case fileList: java.util.List[_] if !fileList.isEmpty =>
            fileList.get(0) match {
              case f: File => Some(f)
              case _ => None
            }
          case _ => None
        }

      } else if (t.isDataFlavorSupported(DataFlavor.stringFlavor)) {
        Some(new File(t.getTransferData(DataFlavor.stringFlavor).asInstanceOf[String]))
      } else {
        None
      }

      newPath.foreach(setFileAndDispatchEvent)
      newPath.isDefined
    }

    override def getSourceActions(c: JComponent): Int = TransferHandler.COPY

    protected override def createTransferable(c: JComponent): Transferable = file match {
      case Some(f) => new PathTransferable(f)
      case _ => null
    }

    protected override def exportDone(source: JComponent, data: Transferable, action: Int) {}

    override def canImport(c: JComponent, flavors: Array[DataFlavor]): Boolean = {
      {
        var i: Int = 0
        while (i < flavors.length) {
          {
            {
              var j: Int = 0
              while (j < supportedFlavors.length) {
                {
                  if (flavors(i) == supportedFlavors(j)) return true
                }
                ({
                  j += 1; j - 1
                })
              }
            }
          }
          ({
            i += 1; i - 1
          })
        }
      }
      false
    }
  }
}